### DISCUSSION TITLE
Why does having a matching .py file on my PYTHONPATH cause gui_scripts to do nothing?

### DISCUSSION THREAD

**Discussion by PySimpleGUI** [pypa/setuptools#3694] [Q&A] [answered]
I use the gui_scripts entry points heavily in PySimpleGUI.

My setup.py has this section:
```python
    entry_points={
        'gui_scripts': [
            'psgissue=PySimpleGUI.PySimpleGUI:main_open_github_issue',
            'psgmain=PySimpleGUI.PySimpleGUI:_main_entry_point',
            'psgupgrade=PySimpleGUI.PySimpleGUI:_upgrade_entry_point',
            'psghelp=PySimpleGUI.PySimpleGUI:main_sdk_help',
            'psgver=PySimpleGUI.PySimpleGUI:main_get_debug_data',
            'psgsettings=PySimpleGUI.PySimpleGUI:main_global_pysimplegui_settings',
        ],},
```

This feature of setuptools is incredibly powerful and has enabled releasing applications to PyPI.  Here is what running `psgmain` from the command line looks like.

![cmd_ZbSJnFoLc1](https://user-images.githubusercontent.com/46163555/202857770-fa76d91f-dd04-4612-beb4-03c4eb41b1f5.gif)

## The Question

The problem I have is that I use a "local" copy of my PySimpleGUI.py file as well as the ones found in each of the Python installations.

I've got .EXE files for each of the installs of PySimpleGUI for each of the Python interpreters I have on my machine.

![image](https://user-images.githubusercontent.com/46163555/202857994-1b03f2fc-d9c6-43c7-854c-90ab882333b4.png)

**They all work great**, except for under one condition.  If there is a PySimpleGUI.py file found on my `PYTHONPATH`, then running `psgmain` and other commands has no effect.  They show no window.  They don't produce an error.  They simply return as if a window was started but none are.

The way I get around this is that I run a little script that renames my PySimpleGUI.py to PySimpleGUI_.py and then I'm able to run these commands.

**Comment by abravalheri** [comment, marked as answer]
I believe this is expected due to the way Python works: in a regular invocation with no special flags, Python will always place the CWD in front of the other directories in `sys.path`.

`pip` will create a wrapper for the entry-points that imports the package according to the current state of `sys.path`.

If the local `PySimpleGUI.py` has problems, then the script will also have problems.


--- EDIT ---

See details in https://github.com/pypa/setuptools/discussions/3694#discussioncomment-4209840

**Comment by PySimpleGUI** [comment]
> If the local PySimpleGUI.py has problems, then the script will also have problems.

The local version has no problems in that it runs as it should, has the same entry points, etc.

**Reply by abravalheri** [reply]
Please note that your entry-point description implies in importing `PySimpleGUI.PySimpleGUI`.

This means that the wrapper created by `pip` with the help of `distutils` expects the first entry it can find in `sys.path` to be a directory named `PySimpleGUI` with a `PySimpleGUI.py` script in it or a nested `PySimpleGUI/__init__.py` file.

This implies that a simple `PySimpleGUI.py` in `sys.path` will mess up the imports because it does not have the same Python API as expected from a normal package installation (which in a simplified manner is equivalent to say that the local `PySimpleGUI.py` file has a problem: lack of API compatibility).

---

If that is not the root cause of the problem, I am afraid I will not be able to help you.

If you are stuck investigating this issue, I would suggest doing a `"cat"` to see the contents of one of the `.exe` files. After the binary blob you should be able to see the code of the wrapper script. Probably something like:

```python
# -*- coding: utf-8 -*-
import re
import sys
from PySimpleGUI.PySimpleGUI import _main_entry_point
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(_main_entry_point())
```

Please note that the wrapper scripts are very likely to be generated by `pip` with the help of `distlib` (unless you are using something like `python setup.py install` or an old version of setuptools) and therefore there is not much setuptools can change to help with that...

**Comment by PySimpleGUI** [comment]
Thank you **very much** for the additional information.  This helped tremendously and explains why my local file has an issue as I don't have the exact same structure.  Really appreciate the help!
